# 4.9 Data Visualizations - Part 1

## Contents:
### Import libraries, set path
### Import customers data
### Update header names with appropriate values as needed
### View column info for NaN values
### View overall statistics of table
### Confirm NaN values
### Check for mixed-type columns
### Check for duplicate records
### Replace NaN values in first_name with appropriate text
### Import current Instacart data
### Join df_customers with df_ords_prods_cust¶
# -----------------------------------

### Import libraries and set path

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
# Set path for use in import
path = r'C:\Users\XLT2\CFProjects\2023-04-07 Instacart Basket Analysis'

### Import customers data

In [3]:
# Import customers data set
df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

In [4]:
# Confirm good import
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [5]:
# Confirm df shape
df_customers.shape

(206209, 10)

### Update header names with appropriate values as needed

In [7]:
# Create dictionary of headers to change
dict = {'First Name': 'first_name',
        'Surnam': 'last_name',
        'Gender': 'gender',
        'STATE': 'state',
        'Age': 'age',
        'n_dependants': 'num_of_dependents',
        'fam_status': 'marital_status'}

In [8]:
# Rename all column headers
df_customers.rename(columns = dict, inplace = True)

In [9]:
# Confirm headers were changed and inspect data
df_customers.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   num_of_dependents marital_status  income  
0                  3        married  165665  
1                  0         single   59285  
2                  2        married   99568  
3                  0         single   42049  
4                  1        married   40374

### View column info for NaN values

In [10]:
# View table info
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   user_id            206209 non-null  int64 
 1   first_name         194950 non-null  object
 2   last_name          206209 non-null  object
 3   gender             206209 non-null  object
 4   state              206209 non-null  object
 5   age                206209 non-null  int64 
 6   date_joined        206209 non-null  object
 7   num_of_dependents  206209 non-null  int64 
 8   marital_status     206209 non-null  object
 9   income             206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [11]:
# Confirm total row count for NaN test
df_customers.shape

(206209, 10)

### Note on NULL values
##### The first_name does not have the same count as the others indicating it has NULL values

### View overall statistics of table

In [12]:
# View statistics
df_customers.describe()

user_id            age  num_of_dependents         income
count  206209.000000  206209.000000      206209.000000  206209.000000
mean   103105.000000      49.501646           1.499823   94632.852548
std     59527.555167      18.480962           1.118433   42473.786988
min         1.000000      18.000000           0.000000   25903.000000
25%     51553.000000      33.000000           0.000000   59874.000000
50%    103105.000000      49.000000           1.000000   93547.000000
75%    154657.000000      66.000000           3.000000  124244.000000
max    206209.000000      81.000000           3.000000  593901.000000

#### No issues are identified via describe()

### Confirm NaN values

In [13]:
# Check NaN values
df_customers.isnull().sum()

user_id                  0
first_name           11259
last_name                0
gender                   0
state                    0
age                      0
date_joined              0
num_of_dependents        0
marital_status           0
income                   0
dtype: int64

#### first_name contains 11259 NaN values

### Check for mixed-type columns

In [14]:
# Check for mixed-type columns
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)

first_name


#### first_name contains mixed-type values

In [15]:
# Confirm data type for first_name
df_customers['first_name'].dtype

dtype('O')

### Check for duplicate records

In [16]:
# Identify all duplicate records for confirmation
df_customers_dups = df_customers[df_customers.duplicated()]

In [17]:
# Confirm duplicate table
df_customers_dups.head()

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, num_of_dependents, marital_status, income]
Index: []

#### No duplicate records found

### Replace NaN values in first_name with appropriate text

In [18]:
# Replace NaN values in first_name with 'NotProvided'
df_customers['first_name'] = df_customers['first_name'].replace(np.nan, 'NotProvided')

In [19]:
# Check NaN values were resolved
df_customers.isnull().sum()

user_id              0
first_name           0
last_name            0
gender               0
state                0
age                  0
date_joined          0
num_of_dependents    0
marital_status       0
income               0
dtype: int64

#### All NaN values in first_name changed: shows zero NaN values

### Write df_customers to file

In [20]:
# Export df_customers as df_customers.pkl
df_customers.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'customers.pkl'))

### Import current Instacart data

In [21]:
# Import orders_products_w_customer_flags.pkl
df_ords_prods_cust = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_w_customer_flags.pkl'))

In [22]:
# Confirm import
df_ords_prods_cust.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  order_dow  ...  \
0   3139998      138    prior            28          6  ...   
1   1977647      138    prior            30          6  ...   
2    389851      709    prior             2          0  ...   
3    652770      764    prior             1          3  ...   
4   1813452      764    prior             3          4  ...   

     price_range_loc     busiest_day       busiest_days_2  \
0  Mid-range product  Regularly busy  Regularly busy days   
1  Mid-range product  Regularly busy  Regularly busy days   
2  Mid-range product     Busiest day         Busiest days   
3  Mid-range product  Regularly busy      Least busy days   
4  Mid-range product      Least busy      Least busy days   

          busiest_hours max_order      loyalty_flag avg_spend spending_flag  \
0         Busiest hours        32  Regular customer  6.935811   Low spender   
1  Regularly busy hours        32  Regular customer  6.935811   Low spender   
2  Regularly busy hours         5      New customer  7.930208   Low spender   
3  Regularly busy hours         3      New customer  4.972414   Low spender   
4  Regularly busy hours         3      New customer  4.972414   Low spender   

  frequency_median     frequency_flag  
0              8.0  Frequent customer  
1              8.0  Frequent customer  
2              6.0  Frequent customer  
3              9.0  Frequent customer  
4              9.0  Frequent customer  

[5 rows x 25 columns]

In [23]:
# Confirm columns in customer needed for join
df_customers.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   num_of_dependents marital_status  income  
0                  3        married  165665  
1                  0         single   59285  
2                  2        married   99568  
3                  0         single   42049  
4                  1        married   40374

In [24]:
# Confir size of customers for check after merge
df_customers.shape

(206209, 10)

In [25]:
df_ords_prods_cust.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  order_dow  ...  \
0   3139998      138    prior            28          6  ...   
1   1977647      138    prior            30          6  ...   
2    389851      709    prior             2          0  ...   
3    652770      764    prior             1          3  ...   
4   1813452      764    prior             3          4  ...   

     price_range_loc     busiest_day       busiest_days_2  \
0  Mid-range product  Regularly busy  Regularly busy days   
1  Mid-range product  Regularly busy  Regularly busy days   
2  Mid-range product     Busiest day         Busiest days   
3  Mid-range product  Regularly busy      Least busy days   
4  Mid-range product      Least busy      Least busy days   

          busiest_hours max_order      loyalty_flag avg_spend spending_flag  \
0         Busiest hours        32  Regular customer  6.935811   Low spender   
1  Regularly busy hours        32  Regular customer  6.935811   Low spender   
2  Regularly busy hours         5      New customer  7.930208   Low spender   
3  Regularly busy hours         3      New customer  4.972414   Low spender   
4  Regularly busy hours         3      New customer  4.972414   Low spender   

  frequency_median     frequency_flag  
0              8.0  Frequent customer  
1              8.0  Frequent customer  
2              6.0  Frequent customer  
3              9.0  Frequent customer  
4              9.0  Frequent customer  

[5 rows x 25 columns]

In [26]:
df_ords_prods_cust.shape

(32404859, 25)

In [27]:
df_ords_prods_cust = df_ords_prods_cust.drop(columns = ['_merge'])

### Join df_customers with df_ords_prods_cust

In [28]:
# Merge customers with ords_prods_cust
df_ords_prods_cust_merged = df_ords_prods_cust.merge(df_customers, on = ['user_id'], how = 'left', indicator = True)

In [29]:
# Confirm merge
df_ords_prods_cust_merged.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  order_dow  ...  first_name  \
0   3139998      138    prior            28          6  ...     Charles   
1   1977647      138    prior            30          6  ...     Charles   
2    389851      709    prior             2          0  ...     Deborah   
3    652770      764    prior             1          3  ...     Heather   
4   1813452      764    prior             3          4  ...     Heather   

   last_name  gender      state age date_joined num_of_dependents  \
0        Cox    Male  Minnesota  81    8/1/2019                 1   
1        Cox    Male  Minnesota  81    8/1/2019                 1   
2      Glass  Female    Vermont  66   6/16/2018                 2   
3      Myers  Female  Wisconsin  40    2/9/2020                 3   
4      Myers  Female  Wisconsin  40    2/9/2020                 3   

  marital_status  income _merge  
0        married   49620   both  
1        married   49620   both  
2        married  158302   both  
3        married   31308   both  
4        married   31308   both  

[5 rows x 34 columns]

In [30]:
df_ords_prods_cust_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 34 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   product_id              int64   
 1   product_name            object  
 2   aisle_id                int64   
 3   department_id           int64   
 4   prices                  float64 
 5   order_id                int64   
 6   user_id                 int64   
 7   eval_set                object  
 8   order_number            int64   
 9   order_dow               int64   
 10  order_hour_of_day       int64   
 11  days_since_prior_order  float64 
 12  add_to_cart_order       int64   
 13  reordered               int64   
 14  price_range_loc         object  
 15  busiest_day             object  
 16  busiest_days_2          object  
 17  busiest_hours           object  
 18  max_order               int64   
 19  loyalty_flag            object  
 20  avg_spend               float64 
 21  spendi

In [31]:
df_ords_prods_cust_merged.shape

(32404859, 34)

In [32]:
# Confirm for full match
df_ords_prods_cust_merged['_merge'].value_counts()

both          32404859
left_only            0
right_only           0
Name: _merge, dtype: int64

#### Row count indicates that all rows from ords_prods_cust_merged retained

### Export ords_prods_merge

In [33]:
# Export orders_products_w_full_customer_info.pkl
df_ords_prods_cust_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_w_full_customer_info.pkl'))

In [34]:
# Import orders_products_w_full_customer_info.pkl for write confirmation
df_temp = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_w_full_customer_info.pkl'))

In [35]:
# Confirm good write
df_temp.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  order_dow  ...  first_name  \
0   3139998      138    prior            28          6  ...     Charles   
1   1977647      138    prior            30          6  ...     Charles   
2    389851      709    prior             2          0  ...     Deborah   
3    652770      764    prior             1          3  ...     Heather   
4   1813452      764    prior             3          4  ...     Heather   

   last_name  gender      state age date_joined num_of_dependents  \
0        Cox    Male  Minnesota  81    8/1/2019                 1   
1        Cox    Male  Minnesota  81    8/1/2019                 1   
2      Glass  Female    Vermont  66   6/16/2018                 2   
3      Myers  Female  Wisconsin  40    2/9/2020                 3   
4      Myers  Female  Wisconsin  40    2/9/2020                 3   

  marital_status  income _merge  
0        married   49620   both  
1        married   49620   both  
2        married  158302   both  
3        married   31308   both  
4        married   31308   both  

[5 rows x 34 columns]